# First Part

objectives: 

1. Scrape wikipedia for toronto boroughs. 
2. Pass data into dataframe
3. Clean data and sort into friendly format

In [1]:
from bs4 import BeautifulSoup

In [2]:
import requests
import pandas as pd

In [3]:
source=requests.get('https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M').text

In [4]:
soup=BeautifulSoup(source, 'lxml')
article=soup.find('table')
print(article.prettify())

<table class="wikitable sortable">
 <tbody>
  <tr>
   <th>
    Postcode
   </th>
   <th>
    Borough
   </th>
   <th>
    Neighbourhood
   </th>
  </tr>
  <tr>
   <td>
    M1A
   </td>
   <td>
    Not assigned
   </td>
   <td>
    Not assigned
   </td>
  </tr>
  <tr>
   <td>
    M2A
   </td>
   <td>
    Not assigned
   </td>
   <td>
    Not assigned
   </td>
  </tr>
  <tr>
   <td>
    M3A
   </td>
   <td>
    <a href="/wiki/North_York" title="North York">
     North York
    </a>
   </td>
   <td>
    <a href="/wiki/Parkwoods" title="Parkwoods">
     Parkwoods
    </a>
   </td>
  </tr>
  <tr>
   <td>
    M4A
   </td>
   <td>
    <a href="/wiki/North_York" title="North York">
     North York
    </a>
   </td>
   <td>
    <a href="/wiki/Victoria_Village" title="Victoria Village">
     Victoria Village
    </a>
   </td>
  </tr>
  <tr>
   <td>
    M5A
   </td>
   <td>
    <a href="/wiki/Downtown_Toronto" title="Downtown Toronto">
     Downtown Toronto
    </a>
   </td>
   <td>
    <a href="

extract the 'useful stuff' from the html

In [5]:
MyTable=soup.find('table', {'class':'wikitable sortable'})
#MyTable

In [6]:
content=MyTable.findAll('tr')
#content

In [7]:
data=[]
for row in content:
    data.append([t.text.strip() for t in row.findAll('td')])

df=pd.DataFrame(data,columns=['PostalCode', 'Borough', 'Neighbourhood'])
df=df[~df['PostalCode'].isnull()]

print(df.head())

  PostalCode           Borough     Neighbourhood
1        M1A      Not assigned      Not assigned
2        M2A      Not assigned      Not assigned
3        M3A        North York         Parkwoods
4        M4A        North York  Victoria Village
5        M5A  Downtown Toronto      Harbourfront


Multiple lines are not usable due to a Borough not being assigned. We will drop these, and group boroughs according to Postal Code. We will also remove duplicates listed in the records.

In [8]:
df.drop(df[df['Borough']=='Not assigned'].index,axis=0,inplace=True)
df.head()

PostalCode           Borough     Neighbourhood
3        M3A        North York         Parkwoods
4        M4A        North York  Victoria Village
5        M5A  Downtown Toronto      Harbourfront
6        M5A  Downtown Toronto       Regent Park
7        M6A        North York  Lawrence Heights

In [9]:
df1=df.reset_index()
df1.head()

index PostalCode           Borough     Neighbourhood
0      3        M3A        North York         Parkwoods
1      4        M4A        North York  Victoria Village
2      5        M5A  Downtown Toronto      Harbourfront
3      6        M5A  Downtown Toronto       Regent Park
4      7        M6A        North York  Lawrence Heights

In [10]:
df2=df1.groupby("PostalCode").agg(lambda x:','.join(x))

Despite having a borough, some don't have a neighborhood. In these cases the neighbourhood will be the borough.

In [11]:
df2.loc[df2['Neighbourhood']=='Not assigned', 'Neighbourhood']=df2.loc[df2['Neighbourhood']=='Not assigned', 'Borough']

In [12]:
df3=df2.reset_index()

In [13]:
df3['Borough']=df3['Borough'].str.replace('nan|[{}\s]','').str.split(',').apply(set).str.join(',').str.strip(',').str.replace(",{2,}",",")
df3.head()

PostalCode      Borough                         Neighbourhood
0        M1B  Scarborough                         Rouge,Malvern
1        M1C  Scarborough  Highland Creek,Rouge Hill,Port Union
2        M1E  Scarborough       Guildwood,Morningside,West Hill
3        M1G  Scarborough                                Woburn
4        M1H  Scarborough                             Cedarbrae

In [14]:
df3.shape

(103, 3)

# Second Part

In [15]:
!pip install geocoder

     |████████████████████████████████| 102kB 15.4MB/s ta 0:00:01


In [17]:
import geocoder



In [19]:
postal_code='M1G'
g=geocoder.arcgis('{}, Toronto, Ontario'.format(postal_code))
print(g.lat, g.lng)

43.768369121000035 -79.21758999999997


In [20]:
posts=[]
for post in df3['PostalCode']:
    g=geocoder.arcgis('{}, Toronto, Ontario'.format(post))
    posts.append((post,g.lat,g.lng))


[('M1B', 43.811525000000074, -79.19551746399998),
 ('M1C', 43.78566500000005, -79.15872457299997),
 ('M1E', 43.76581500000003, -79.17519294699997),
 ('M1G', 43.768369121000035, -79.21758999999997),
 ('M1H', 43.76968799900004, -79.23943999999995),
 ('M1J', 43.74312500000008, -79.23174973599998),
 ('M1K', 43.72627568400003, -79.26362499999993),
 ('M1L', 43.71305350000006, -79.28505499999994),
 ('M1M', 43.724234575000025, -79.22792499999997),
 ('M1N', 43.69677000000007, -79.25996735299998),
 ('M1P', 43.759975000000054, -79.26897418299995),
 ('M1R', 43.750710464000065, -79.30055999999996),
 ('M1S', 43.79394000000008, -79.26797613999997),
 ('M1T', 43.78472500000004, -79.29906603299997),
 ('M1V', 43.81768500000004, -79.28018721399997),
 ('M1W', 43.80088292800008, -79.32073999999994),
 ('M1X', 43.83421500000003, -79.21670085099998),
 ('M2H', 43.80284500000005, -79.35620744999994),
 ('M2J', 43.780970000000025, -79.34781328099996),
 ('M2K', 43.781015000000025, -79.38052867199997),
 ('M2L', 43.7

In [29]:
postaldf=pd.DataFrame(posts)
postaldf.columns=['PostalCode','Latitude','Longitude']
postaldf.head()

PostalCode   Latitude  Longitude
0        M1B  43.811525 -79.195517
1        M1C  43.785665 -79.158725
2        M1E  43.765815 -79.175193
3        M1G  43.768369 -79.217590
4        M1H  43.769688 -79.239440

In [31]:
mergedf=pd.merge(df3,postaldf[['PostalCode','Latitude','Longitude']], on='PostalCode', how='left')
mergedf.head()

PostalCode      Borough                         Neighbourhood   Latitude  \
0        M1B  Scarborough                         Rouge,Malvern  43.811525   
1        M1C  Scarborough  Highland Creek,Rouge Hill,Port Union  43.785665   
2        M1E  Scarborough       Guildwood,Morningside,West Hill  43.765815   
3        M1G  Scarborough                                Woburn  43.768369   
4        M1H  Scarborough                             Cedarbrae  43.769688   

   Longitude  
0 -79.195517  
1 -79.158725  
2 -79.175193  
3 -79.217590  
4 -79.239440

# Third Part

In [33]:
import json # library to handle JSON files

!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

print('Libraries imported.')

Solving environment: done

# All requested packages already installed.

Solving environment: done

## Package Plan ##

  environment location: /opt/conda/envs/Python36

  added / updated specs: 
    - folium=0.5.0


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    folium-0.5.0               |             py_0          45 KB  conda-forge
    altair-3.2.0               |           py36_0         770 KB  conda-forge
    vincent-0.4.4              |             py_1          28 KB  conda-forge
    branca-0.3.1               |             py_0          25 KB  conda-forge
    ------------------------------------------------------------
                                           Total:         868 KB

The following NEW packages will be INSTALLED:

    altair:  3.2.0-py36_0 conda-forge
    branca:  0.3.1-py_0   conda-forge
    folium:  0.5.0-py_0   conda-forge
    vincent: 0.4.4-py_1   conda-forg

In [36]:
latitude=43.811525
longitude=-79.195517

# create map of Toronto using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighbourhood in zip(mergedf['Latitude'], mergedf['Longitude'], mergedf['Borough'], mergedf['Neighbourhood']):
    label = '{}, {}'.format(neighbourhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=2,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

Now connecting to Foursquare to obtain venues in vicinity of coordinates determined for each neighbourhood.

In [38]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighbourhood', 
                  'Neighbourhood Latitude', 
                  'Neighbourhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [41]:
LIMIT = 100
radius = 500
toronto_venues = getNearbyVenues(names=mergedf['Neighbourhood'],
                                   latitudes=mergedf['Latitude'],
                                   longitudes=mergedf['Longitude']
                                  )

Rouge,Malvern
Highland Creek,Rouge Hill,Port Union
Guildwood,Morningside,West Hill
Woburn
Cedarbrae
Scarborough Village
East Birchmount Park,Ionview,Kennedy Park
Clairlea,Golden Mile,Oakridge
Cliffcrest,Cliffside,Scarborough Village West
Birch Cliff,Cliffside West
Dorset Park,Scarborough Town Centre,Wexford Heights
Maryvale,Wexford
Agincourt
Clarks Corners,Sullivan,Tam O'Shanter
Agincourt North,L'Amoreaux East,Milliken,Steeles East
L'Amoreaux West
Upper Rouge
Hillcrest Village
Fairview,Henry Farm,Oriole
Bayview Village
Silver Hills,York Mills
Newtonbrook,Willowdale
Willowdale South
York Mills West
Willowdale West
Parkwoods
Don Mills North
Flemingdon Park,Don Mills South
Bathurst Manor,Downsview North,Wilson Heights
Northwood Park,York University
CFB Toronto,Downsview East
Downsview West
Downsview Central
Downsview Northwest
Victoria Village
Woodbine Gardens,Parkview Hill
Woodbine Heights
The Beaches
Leaside
Thorncliffe Park
East Toronto
The Danforth West,Riverdale
The Beaches West,Indi

In [43]:
toronto_venues.head()

Neighbourhood  Neighbourhood Latitude  \
0  Highland Creek,Rouge Hill,Port Union               43.785665   
1       Guildwood,Morningside,West Hill               43.765815   
2       Guildwood,Morningside,West Hill               43.765815   
3       Guildwood,Morningside,West Hill               43.765815   
4                                Woburn               43.768369   

   Neighbourhood Longitude                        Venue  Venue Latitude  \
0               -79.158725        Royal Canadian Legion       43.782533   
1               -79.175193     Homestead Roofing Repair       43.765140   
2               -79.175193  Heron Park Community Centre       43.768867   
3               -79.175193                   Heron Park       43.769327   
4               -79.217590                    Starbucks       43.770037   

   Venue Longitude              Venue Category  
0       -79.163085                         Bar  
1       -79.178663  Construction & Landscaping  
2       -79.176958        Gym / Fitness Center  
3       -79.177201                        Park  
4       -79.221156                 Coffee Shop

In [42]:
toronto_venues.groupby('Neighbourhood').count()

Neighbourhood Latitude  \
Neighbourhood                                                                
Adelaide,King,Richmond                                                 100   
Agincourt                                                               13   
Agincourt North,L'Amoreaux East,Milliken,Steele...                       2   
Albion Gardens,Beaumond Heights,Humbergate,Jame...                      13   
Alderwood,Long Branch                                                    4   
Bayview Village                                                          5   
Bedford Park,Lawrence Manor East                                        20   
Berczy Park                                                             62   
Birch Cliff,Cliffside West                                               6   
Bloordale Gardens,Eringate,Markland Wood,Old Bu...                       6   
Brockton,Exhibition Place,Parkdale Village                              68   
Business Reply Mail Processing Centre 969 Eastern                      100   
CFB Toronto,Downsview East                                               4   
CN Tower,Bathurst Quay,Island airport,Harbourfr...                      70   
Cabbagetown,St. James Town                                              38   
Caledonia-Fairbanks                                                     10   
Canada Post Gateway Processing Centre                                  100   
Cedarbrae                                                                2   
Central Bay Street                                                      97   
Chinatown,Grange Park,Kensington Market                                 80   
Christie                                                                11   
Church and Wellesley                                                    82   
Clairlea,Golden Mile,Oakridge                                           11   
Clarks Corners,Sullivan,Tam O'Shanter                                   12   
Cliffcrest,Cliffside,Scarborough Village West                            6   
Cloverdale,Islington,Martin Grove,Princess Gard...                       2   
Commerce Court,Victoria Hotel                                          100   
Davisville                                                              29   
Davisville North                                                         7   
Deer Park,Forest Hill SE,Rathnelly,South Hill,S...                       6   
...                                                                    ...   
North Toronto West                                                       4   
Northwest                                                                4   
Northwood Park,York University                                          13   
Parkdale,Roncesvalles                                                   55   
Parkwoods                                                                3   
Queen's Park                                                            51   
Rosedale                                                                 4   
Roselawn                                                                 1   
Runnymede,Swansea                                                       41   
Ryerson,Garden District                                                100   
Scarborough Village                                                      4   
St. James Town                                                         100   
Stn A PO Boxes 25 The Esplanade                                        100   
Studio District                                                         50   
The Annex,North Midtown,Yorkville                                       24   
The Beaches                                                              4   
The Beaches West,India Bazaar                                           19   
The Danforth West,Riverdale                                              4   
The Junction North,Runnymede                                             5

In [45]:
print('There are {} unique categories.'.format(len(toronto_venues['Venue Category'].unique())))

There are 260 unique categories.


Now inserting OneHot coding for the 260 unique categories of venue.

In [47]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighbourhood'] = toronto_venues['Neighbourhood'] 

# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot.head()

Neighbourhood  Afghan Restaurant  Airport  \
0  Highland Creek,Rouge Hill,Port Union                  0        0   
1       Guildwood,Morningside,West Hill                  0        0   
2       Guildwood,Morningside,West Hill                  0        0   
3       Guildwood,Morningside,West Hill                  0        0   
4                                Woburn                  0        0   

   American Restaurant  Antique Shop  Art Gallery  Arts & Crafts Store  \
0                    0             0            0                    0   
1                    0             0            0                    0   
2                    0             0            0                    0   
3                    0             0            0                    0   
4                    0             0            0                    0   

   Asian Restaurant  Athletics & Sports  Auto Dealership  ...  Train Station  \
0                 0                   0                0  ...              0   
1                 0                   0                0  ...              0   
2                 0                   0                0  ...              0   
3                 0                   0                0  ...              0   
4                 0                   0                0  ...              0   

   Tram Station  Vegetarian / Vegan Restaurant  Video Game Store  \
0             0                              0                 0   
1             0                              0                 0   
2             0                              0                 0   
3             0                              0                 0   
4             0                              0                 0   

   Vietnamese Restaurant  Waterfall  Wine Bar  Wings Joint  Women's Store  \
0                      0          0         0            0              0   
1                      0          0         0            0              0   
2                      0          0         0            0              0   
3                      0          0         0            0              0   
4                      0          0         0            0              0   

   Yoga Studio  
0            0  
1            0  
2            0  
3            0  
4            0  

[5 rows x 261 columns]

In [48]:
toronto_grouped = toronto_onehot.groupby('Neighbourhood').mean().reset_index()
toronto_grouped

Neighbourhood  Afghan Restaurant  \
0                              Adelaide,King,Richmond           0.000000   
1                                           Agincourt           0.000000   
2   Agincourt North,L'Amoreaux East,Milliken,Steel...           0.000000   
3   Albion Gardens,Beaumond Heights,Humbergate,Jam...           0.000000   
4                               Alderwood,Long Branch           0.000000   
5                                     Bayview Village           0.000000   
6                    Bedford Park,Lawrence Manor East           0.000000   
7                                         Berczy Park           0.000000   
8                          Birch Cliff,Cliffside West           0.000000   
9   Bloordale Gardens,Eringate,Markland Wood,Old B...           0.000000   
10         Brockton,Exhibition Place,Parkdale Village           0.000000   
11  Business Reply Mail Processing Centre 969 Eastern           0.000000   
12                         CFB Toronto,Downsview East           0.000000   
13  CN Tower,Bathurst Quay,Island airport,Harbourf...           0.000000   
14                         Cabbagetown,St. James Town           0.000000   
15                                Caledonia-Fairbanks           0.000000   
16              Canada Post Gateway Processing Centre           0.000000   
17                                          Cedarbrae           0.000000   
18                                 Central Bay Street           0.000000   
19            Chinatown,Grange Park,Kensington Market           0.000000   
20                                           Christie           0.000000   
21                               Church and Wellesley           0.012195   
22                      Clairlea,Golden Mile,Oakridge           0.000000   
23              Clarks Corners,Sullivan,Tam O'Shanter           0.000000   
24      Cliffcrest,Cliffside,Scarborough Village West           0.000000   
25  Cloverdale,Islington,Martin Grove,Princess Gar...           0.000000   
26                      Commerce Court,Victoria Hotel           0.000000   
27                                         Davisville           0.000000   
28                                   Davisville North           0.000000   
29  Deer Park,Forest Hill SE,Rathnelly,South Hill,...           0.000000   
..                                                ...                ...   
69                                 North Toronto West           0.000000   
70                                          Northwest           0.000000   
71                     Northwood Park,York University           0.000000   
72                              Parkdale,Roncesvalles           0.000000   
73                                          Parkwoods           0.000000   
74                                       Queen's Park           0.000000   
75                                           Rosedale           0.000000   
76                                           Roselawn           0.000000   
77                                  Runnymede,Swansea           0.000000   
78                            Ryerson,Garden District           0.000000   
79                                Scarborough Village           0.000000   
80                                     St. James Town           0.000000   
81                    Stn A PO Boxes 25 The Esplanade           0.000000   
82                                    Studio District           0.000000   
83                  The Annex,North Midtown,Yorkville           0.000000   
84                                        The Beaches           0.000000   
85                      The Beaches West,India Bazaar           0.000000   
86                        The Danforth West,Riverdale           0.000000   
87                       The Junction North,Runnymede           0.000000   
88        The Kingsway,Montgomery Road,Old Mill North           0.000000   
89                                   Thorncliffe Park           0.000000   
90          

In [52]:
import numpy as np

In [50]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

With the top 10 venues, as follows, we will be able to run the clustering algorithm

In [70]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighbourhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighbourhood'] = toronto_grouped['Neighbourhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighbourhood 1st Most Common Venue  \
0                             Adelaide,King,Richmond           Coffee Shop   
1                                          Agincourt         Shopping Mall   
2  Agincourt North,L'Amoreaux East,Milliken,Steel...              Pharmacy   
3  Albion Gardens,Beaumond Heights,Humbergate,Jam...         Grocery Store   
4                              Alderwood,Long Branch     Convenience Store   

  2nd Most Common Venue        3rd Most Common Venue 4th Most Common Venue  \
0                  Café                        Hotel            Steakhouse   
1           Supermarket                Grocery Store      Sushi Restaurant   
2           Yoga Studio  Eastern European Restaurant           Flea Market   
3          Liquor Store               Sandwich Place   Japanese Restaurant   
4        Sandwich Place                          Gym                   Pub   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0   Japanese Restaurant                   Bar   American Restaurant   
1                Bakery        Discount Store       Badminton Court   
2           Fish Market     Fish & Chips Shop   Filipino Restaurant   
3   Fried Chicken Joint           Coffee Shop            Beer Store   
4    Falafel Restaurant     Electronics Store     Elementary School   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0                Bakery          Burger Joint         Breakfast Spot  
1                  Pool    Chinese Restaurant  Vietnamese Restaurant  
2                 Field  Fast Food Restaurant         Farmers Market  
3           Pizza Place                  Park   Fast Food Restaurant  
4  Ethiopian Restaurant           Event Space                   Farm

In [71]:
# set number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighbourhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([3, 3, 0, 3, 3, 1, 3, 3, 3, 3], dtype=int32)

In [72]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = mergedf

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighbourhood'), on='Neighbourhood')



PostalCode      Borough                         Neighbourhood   Latitude  \
0        M1B  Scarborough                         Rouge,Malvern  43.811525   
1        M1C  Scarborough  Highland Creek,Rouge Hill,Port Union  43.785665   
2        M1E  Scarborough       Guildwood,Morningside,West Hill  43.765815   
3        M1G  Scarborough                                Woburn  43.768369   
4        M1H  Scarborough                             Cedarbrae  43.769688   

   Longitude  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0 -79.195517             NaN                   NaN                   NaN   
1 -79.158725             4.0                   Bar           Yoga Studio   
2 -79.175193             1.0                  Park  Gym / Fitness Center   
3 -79.217590             1.0           Coffee Shop      Business Service   
4 -79.239440             3.0            Playground                 Trail   

        3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0                         NaN                   NaN                   NaN   
1                        Food           Flower Shop           Flea Market   
2  Construction & Landscaping     Elementary School  Ethiopian Restaurant   
3           Korean Restaurant                  Park  Fast Food Restaurant   
4                        Farm     Electronics Store     Elementary School   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0                   NaN                   NaN                   NaN   
1           Fish Market     Fish & Chips Shop   Filipino Restaurant   
2           Event Space    Falafel Restaurant                  Farm   
3           Event Space    Falafel Restaurant                  Farm   
4  Ethiopian Restaurant           Event Space    Falafel Restaurant   

  9th Most Common Venue 10th Most Common Venue  
0                   NaN                    NaN  
1                 Field   Fast Food Restaurant  
2        Farmers Market            Yoga Studio  
3        Farmers Market            Yoga Studio  
4        Farmers Market    Dumpling Restaurant

It is necessary to drop the few rows that returned no venues, and then to transform the data type of the cluster labels to int.

In [84]:
toronto_merged=toronto_merged.dropna()

In [76]:
convert_dict={'Cluster Labels':int}
toronto_merged=toronto_merged.astype(convert_dict)

In [77]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighbourhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

Below are the clusters to visualize how they are composed. Cluster 1 seems to be the most eclectic, while 3 seems to have a large variety of restaurants.

In [78]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 0, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
14  Scarborough               0              Pharmacy           Yoga Studio   

          3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
14  Eastern European Restaurant           Flea Market           Fish Market   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
14     Fish & Chips Shop   Filipino Restaurant                 Field   

   9th Most Common Venue 10th Most Common Venue  
14  Fast Food Restaurant         Farmers Market

In [79]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 1, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels       1st Most Common Venue  \
2       Scarborough               1                        Park   
3       Scarborough               1                 Coffee Shop   
19        NorthYork               1  Construction & Landscaping   
23        NorthYork               1           Convenience Store   
25        NorthYork               1           Food & Drink Shop   
26        NorthYork               1                        Park   
30        NorthYork               1                        Park   
34        NorthYork               1                        Park   
40         EastYork               1                         Bar   
41      EastToronto               1                    Bus Line   
46   CentralToronto               1                  Playground   
50  DowntownToronto               1                  Playground   
59  DowntownToronto               1                        Park   
64   CentralToronto               1                        Park   
73             York               1                Hockey Arena   
74             York               1                        Park   
79        NorthYork               1            Basketball Court   
82      WestToronto               1           Convenience Store   
90        Etobicoke               1                        Park   
97        NorthYork               1                 Coffee Shop   

                       2nd Most Common Venue        3rd Most Common Venue  \
2                       Gym / Fitness Center   Construction & Landscaping   
3                           Business Service            Korean Restaurant   
19                        Golf Driving Range                      Dog Run   
23                                      Park                         Bank   
25                                  Bus Stop                         Park   
26                              Burger Joint                  Coffee Shop   
30                                   Airport                   Food Court   
34                                Food Stand                Grocery Store   
40                            Farmers Market           Italian Restaurant   
41                             Grocery Store                         Park   
46                                  Gym Pool                       Garden   
50                               Candy Store                Grocery Store   
59                                      Pier              Harbor / Marina   
64                                 Waterfall                       Lawyer   
73                                     Field                  Coffee Shop   
74                                    Market                   Beer Store   
79                                    Bakery                         Park   
82  Residential Building (Apartment / Condo)                         Park   
90                               Yoga Studio  Eastern European Restaurant   
97                                 Nightclub                         Park   

   4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
2      Elementary School  Ethiopian Restaurant           Event Space   
3                   Park  Fast Food Restaurant           Event Space   
19                  Park                 Trail           Yoga Studio   
23             Speakeasy          Dance Studio           Flower Shop   
25     Electronics Store           Flea Market           Fish Market   
26  Fast Food Restaurant  Ethiopian Restaurant           Event Space   
30           Coffee Shop  Fast Food Restaurant  Ethiopian Restaurant   
34           Yoga Studio     Elementary School  Ethiopian Restaurant   
40                  Park     Elementary School  Ethiopian Restaurant   
41        Discount Store           Yoga Studio        Farmers Market   
46                  Park   Filipino Restaurant                 Field   
50                  Park   Filipino Restaurant                 Field   
59  Fast Food Restaurant     Elementary School  Et

In [80]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 2, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
32  NorthYork               2          Home Service     Food & Drink Shop   
94  Etobicoke               2          Home Service   Filipino Restaurant   
96  NorthYork               2          Home Service     Food & Drink Shop   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
32           Flower Shop           Flea Market           Fish Market   
94     Food & Drink Shop           Flower Shop           Flea Market   
96           Flower Shop           Flea Market           Fish Market   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
32     Fish & Chips Shop   Filipino Restaurant                 Field   
94           Fish Market     Fish & Chips Shop                 Field   
96     Fish & Chips Shop   Filipino Restaurant                 Field   

   9th Most Common Venue 10th Most Common Venue  
32  Fast Food Restaurant         Farmers Market  
94  Fast Food Restaurant         Farmers Market  
96  Fast Food Restaurant         Farmers Market

In [81]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 3, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels   1st Most Common Venue  \
4        Scarborough               3              Playground   
5        Scarborough               3              Restaurant   
6        Scarborough               3      Chinese Restaurant   
7        Scarborough               3            Intersection   
8        Scarborough               3          Sandwich Place   
9        Scarborough               3                Gym Pool   
10       Scarborough               3     Japanese Restaurant   
11       Scarborough               3            Intersection   
12       Scarborough               3           Shopping Mall   
13       Scarborough               3                Pharmacy   
15       Scarborough               3    Fast Food Restaurant   
17         NorthYork               3                 Dog Run   
18         NorthYork               3          Clothing Store   
21         NorthYork               3                    Café   
22         NorthYork               3             Coffee Shop   
24         NorthYork               3            Intersection   
27         NorthYork               3             Supermarket   
29         NorthYork               3  Furniture / Home Store   
31         NorthYork               3       Convenience Store   
33         NorthYork               3           Grocery Store   
35          EastYork               3             Pizza Place   
36          EastYork               3                Pharmacy   
37       EastToronto               3       Health Food Store   
38          EastYork               3             Coffee Shop   
39          EastYork               3             Yoga Studio   
42       EastToronto               3          Sandwich Place   
43       EastToronto               3                   Diner   
44    CentralToronto               3                Bus Line   
45    CentralToronto               3          Clothing Store   
47    CentralToronto               3            Dessert Shop   
..               ...             ...                     ...   
65    CentralToronto               3          Sandwich Place   
66   DowntownToronto               3                    Café   
67   DowntownToronto               3                    Café   
68   DowntownToronto               3             Coffee Shop   
69   DowntownToronto               3             Coffee Shop   
70   DowntownToronto               3             Coffee Shop   
71         NorthYork               3          Clothing Store   
72         NorthYork               3             Pizza Place   
75   DowntownToronto               3                    Café   
76       WestToronto               3                    Park   
77       WestToronto               3                     Bar   
78       WestToronto               3             Coffee Shop   
80              York               3              Playground   
81              York               3                 Brewery   
83       WestToronto               3             Coffee Shop   
84       WestToronto               3                    Café   
85       Queen'sPark               3             Coffee Shop   
86       Mississauga               3             Coffee Shop   
87       EastToronto               3             Coffee Shop   
88         Etobicoke               3       Convenience Store   
89         Etobicoke               3       Convenience Store   
91         Etobicoke               3      Chinese Restaurant   
92         Etobicoke               3           Burrito Place   
93         Etobicoke               3                Pharmacy   
95         Etobicoke               3            Carpet Store   
98              York               3             Pizza Place   
99         Etobicoke               3             Pizza Place   
100        Etobicoke               3             Pizza Place   
101        Etobicoke               3           Grocery Store   
102        Etobicoke               3          Sandwich Place   

                        2nd Most Common Venu

In [82]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 4, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
1  Scarborough               4                   Bar           Yoga Studio   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
1                  Food           Flower Shop           Flea Market   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
1           Fish Market     Fish & Chips Shop   Filipino Restaurant   

  9th Most Common Venue 10th Most Common Venue  
1                 Field   Fast Food Restaurant